### PR TITLE
Add TChannel to IChannelFactory, and related cleanups

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Add TChannel type parameter to IChannelFactory, and a few related cleanups.

Changes split out from https://github.com/microsoft/FluidFramework/pull/19717

## Breaking Changes

Add TChannel type parameter (which defaults to IFluidLoadable) to IChannelFactory. When left at its default this preserves the old behavior, however packages exporting IChannelFactory will now reference IFluidLoadable if not providing a different parameter and thus will implicitly depend on @fluidframework/core-interfaces

SharedCounter.create now returns ISharedCounter instead of SharedCounter

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/framework/fluid-static/src/types.ts [hunk 1]
```diff
@@ -31,8 +31,8 @@ export type LoadableObjectClassRecord = Record<string, LoadableObjectClass<any>>
  * @public
  */
 export type LoadableObjectClass<T extends IFluidLoadable> =
-	| DataObjectClass<T>
-	| SharedObjectClass<T>;
+	| SharedObjectClass<T>
```
Comment (Josmithr) [inline packages/framework/fluid-static/src/types.ts:34]: Is this change necessary?

File: packages/runtime/datastore-definitions/src/channel.ts [hunk 2]
```diff
@@ -262,7 +262,7 @@ export interface IChannelServices {
  * the collaborating clients will need to have access to a factory that can produce the `SharedMap` object.
  * @public
  */
-export interface IChannelFactory {
+export interface IChannelFactory<out TChannel extends IFluidLoadable = IFluidLoadable> {
```
Comment (Abe27342) [inline packages/runtime/datastore-definitions/src/channel.ts:265]: why constrain to and default as `IFluidLoadable` rather than `IChannel`? Seems like it would give simpler types.

File: packages/framework/fluid-static/src/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/framework/fluid-static/src/types.ts:34]: Yes. Swapping the order here prevents one of the shared directory using tests from failing to infer T properly and not compiling. Putting shared object first, which is the more complex and more commonly used type seems to improve type inference.

File: packages/runtime/datastore-definitions/src/channel.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/runtime/datastore-definitions/src/channel.ts:265]: I was unable to make the type inference around isDataObject work correctly when I did that (its what I tried first, and I tried it again when making this PR). A lot of code is generic over shared object and data objects, and parameterized by a IFluidLoadable: making shared objects require I channel seems like it should work, but you end up needing an extends clause or another intersection in the LoadableObject union, and that screws up narrowing. Its likley possible to make work, but I tried and failed twice, so I went with this which works fine and is a smaller delta.

File: packages/dds/counter/api-report/counter.api.md [hunk 3]
```diff
@@ -31,8 +31,8 @@ export interface ISharedCounterEvents extends ISharedObjectEvents {
 export class SharedCounter extends SharedObject<ISharedCounterEvents> implements ISharedCounter {
     constructor(id: string, runtime: IFluidDataStoreRuntime, attributes: IChannelAttributes);
     protected applyStashedOp(op: unknown): void;
-    static create(runtime: IFluidDataStoreRuntime, id?: string): SharedCounter;
-    static getFactory(): IChannelFactory;
+    static create(runtime: IFluidDataStoreRuntime, id?: string): ISharedCounter;
```
Comment (alexvy86) [inline packages/dds/counter/api-report/counter.api.md:34]: This looks like it might need a call out as a breaking change?

Comment (CraigMacomber) [reply packages/dds/counter/api-report/counter.api.md:34]: Done. As this API is only alpha, its not a huge deal, but you are correct it should be noted!

File: packages/framework/fluid-static/src/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/framework/fluid-static/src/types.ts:34]: Might be worth adding a comment noting that?

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/framework/fluid-static/src/types.ts:34]: for reference the error you get is:

@fluidframework/tinylicious-client: error during command 'tsc --project ./src/test/tsconfig.json' (exit code 2)
@fluidframework/tinylicious-client: src/test/TinyliciousClient.spec.ts:251:42 - error TS2345: Argument of type 'typeof SharedDirectory' is not assignable to parameter of type 'LoadableObjectClass<SharedDirectory>'.
@fluidframework/tinylicious-client:   Type 'typeof SharedDirectory' is not assignable to type 'DataObjectClass<SharedDirectory>'.
@fluidframework/tinylicious-client:     Property 'factory' is missing in type 'typeof SharedDirectory' but required in type '{ readonly factory: { IFluidDataStoreFactory: ...; }; }'.
@fluidframework/tinylicious-client: 
@fluidframework/tinylicious-client: 251   const newPair = await container.create(SharedDirectory);
@fluidframework/tinylicious-client:                                              ~~~~~~~~~~~~~~~
@fluidframework/tinylicious-client: 
@fluidframework/tinylicious-client:   ../../framework/fluid-static/dist/types.d.ts:33:14
@fluidframework/tinylicious-client:     33     readonly factory: {
@fluidframework/tinylicious-client:                     ~~~~~~~
@fluidframework/tinylicious-client:     'factory' is declared here.
@fluidframework/tinylicious-client: 
@fluidframework/tinylicious-client: 
@fluidframework/tinylicious-client: Found 1 error in src/test/TinyliciousClient.spec.ts:251

And getting that requires combining this PR's changes with the ones in https://github.com/microsoft/FluidFramework/pull/19717

I'll add a note to the code.

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/framework/fluid-static/src/types.ts:40]: I had added an explication to the comment via privateRemarks.

If you ewant an interactive example one is included in https://www.typescriptlang.org/play?ts=5.3.3#code/PTAEEkBdQWwQ0pApgJwM6gO4AsCWBjbUAexQBNVRJslQBXAO12IY1zYYC4AoAb29CDQ+ADZw0GcAwBmqOACMRtfkNWgADnUUFQ8gISd6DANYNimBgG4BawZu35QaSAh1xDDOjHmprqgL7cNoKQAJ7qtACCADwAKgB8oAC8oCq2woaMpuZWwaD+oABkoAAUDEiYpQCUyYmxVdZ5uAzIKNJw+LQAQnGJabbuoLF+QoFBqmERoACqTCwAjL3JoDEJoAA+oD0JIyHhtLPMDABMSynbiZur8Y2q0oz4kEegABq9JcTyAFZIjwDCYgkhkOC16VUMADdiLgyKlAncHk8WK9TgkPt9fpAAeI0MC5icwZDobDeGNVC8SlJZCgFEoGnkXsdKTI5IokPSxiAINB4IhUBgcAQiKQKCgqDRQM1WmhMUcMHAULRmjx+JMossAEpISB0FAMWL7aIlEo1JKJeY1YrG02JTzeVBVG7cLmCAB6AH5uGqtprtbr9YbrbVQHafChLaUTcGLU6XaAPdx-EA

File: packages/runtime/datastore-definitions/src/channel.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/runtime/datastore-definitions/src/channel.ts:265]: My experiments with this were before I swapped the order of the LoadableObjectClass union. Let me see if it works better now that that has been swapped.

Comment (anthony-murphy) [reply packages/runtime/datastore-definitions/src/channel.ts:265]: i got it to work with a single cast in RootDataObject.create. its necessary due to limitations of type narrowing on generics:
https://github.com/microsoft/FluidFramework/compare/main...anthony-murphy:prototype-ichannel-in-factory

The end typing is also a bit wonky, as i went with interfaces in the factories, which results in union of class and interface due to the dependency of contianer schema on class. this could be resolve by just using the class as the generic parameter to the factory. it may also be possible to parametrize the factory itself, so it works with either classes or interfaces to keep the ability to more away from classes more easily supported.

Comment (CraigMacomber) [reply packages/runtime/datastore-definitions/src/channel.ts:265]: I tried integrating that change into my [AzureFactory branch](https://github.com/microsoft/FluidFramework/pull/19717), and it broke tons of tests using container schema.

The issue is that if:
```typescript
export type LoadableObjectClass<T extends IFluidLoadable = IFluidLoadable> =
	| (T extends IChannel ? SharedObjectClass<T> : never)
	| DataObjectClass<T>;
```

Then `LoadableObjectClass<IFluidLoadable>` simplifies to DataObjectClass<IFluidLoadable> which means that SharedObjectClass no longer can be used in ContainerSchema.

In short, that version of LoadableObjectClass fails to be covariant with respect to T, which breaks a lot of things including shared objects in container schema.

An interesting feature of the approach I took (`& IChannel` on the return ) is that it now becomes optional for interfaces like ITree to extend IChannel, so we could actually hide (in the container schema based API) the fact that trees expose things like "getAttachSummary", "summarize", "getGCData" if we want to.

File: packages/framework/fluid-static/src/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/framework/fluid-static/src/types.ts:41]: This behavior of unions appears to be a TS bug, and a relevant issue has been linked

File: packages/runtime/datastore-definitions/src/channel.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/runtime/datastore-definitions/src/channel.ts:265]: why make that change at all, can't it remain: 
`export type LoadableObjectClass<T extends IFluidLoadable>`

Comment (CraigMacomber) [reply packages/runtime/datastore-definitions/src/channel.ts:272]: My above comment was in response to Abram's proposal, but @anthony-murphy 's approach:

```typescript
export type LoadableObjectClass<T extends IFluidLoadable> = T extends IChannel
	? SharedObjectClass<T>
	: DataObjectClass<T>;
```

Is one I tried previously, and actually has the same issue where LoadableObjectClass<IFluidLoadable> simplifies to DataObjectClass which means that SharedObjectClass no longer can be used in ContainerSchema when combined with the changes in my [AzureFactory branch](https://github.com/microsoft/FluidFramework/pull/19717).

This change also has the issue that it becomes impossible to properly support a data-object that returns a type which happens to implement IChannel: the old code, my changes and Abram's suggestions all allowed that case. THis change results in my adjusted

```typescript
export function isDataObjectClass<T extends IFluidLoadable>(
	obj: LoadableObjectClass<T>,
): obj is InternalDataObjectClass<T>;
```
failing to compile, and needing to be changed to:
```typescript
export function isDataObjectClass<T extends IFluidLoadable>(
	obj: LoadableObjectClass<T> | DataObjectClass<T>,
): obj is InternalDataObjectClass<T>;
```

If we are willing to keep the use of `any` in ContainerSchema (which results in untyped container schema returning DDSs and shared objects as `any` instead of as `IFluidLoadable` as is remove in my other branch as a follow-up to this work) then I suspect we could make this work, but that seems worse than what I have proposed which avoids the cast, avoids the use of any, and only looks a little funny if you read the implementation of IChannelFactory, but nothing about its unsafe from a type perspective.

The current implementation of InitialObjects does produce `any` with [Anthony](https://github.com/anthony-murphy)'s version, but the one on my branch does work around such issues. Thus it's possible to work around the lack of covariance, but evert user of the container schema type would need to do so, and forgetting to do so will produce `any`  and the workaround is messy and a bit fragile.

I think factory style APIs should maintain covariant and thus be able to avoid `any`. My approach accomplishes this while the other suggestions here do not.

I have moved some of changes (removal of any) from my other branch into this one which benefit from the covariance and added a private doc comment about this design choice.

Comment (Abe27342) [reply packages/runtime/datastore-definitions/src/channel.ts:272]: I still think my first proposal builds and solves things for the encapsulated model in isolation. But I didn't realize your other changes actually do introduce typesafety to the azure APIs, at which point I agree the end state is better--seems reasonable to me.

**Review by Abe27342 — APPROVED** [review]

File: packages/dds/map/src/directory.ts [hunk 4]
```diff
@@ -467,16 +467,16 @@ export class SharedDirectory
 	 * @param id - Optional name of the shared directory
 	 * @returns Newly create shared directory (but not attached yet)
 	 */
-	public static create(runtime: IFluidDataStoreRuntime, id?: string): SharedDirectory {
-		return runtime.createChannel(id, DirectoryFactory.Type) as SharedDirectory;
+	public static create(runtime: IFluidDataStoreRuntime, id?: string): ISharedDirectory {
+		return runtime.createChannel(id, DirectoryFactory.Type) as ISharedDirectory;
```
Comment (tyler-cai-microsoft) [inline packages/dds/map/src/directory.ts:471]: I know this may be out of scope, I'm liking this PR already. I wonder if we could add types to `ISharedDirectory<T>` and `ISharedMap<T>`

That would be awesome :).

**Review by Josmithr — APPROVED** [review]
Devtools changes look fine to me.

File: packages/dds/map/src/directory.ts (hunk 4 above)
Comment (CraigMacomber) [reply packages/dds/map/src/directory.ts:471]: There are complexities with doing that since that T might want to be changed over time, which brings in compatibility issues. SharedTree is building quite a lot of stuff to properly support something like that, but I think maybe applying the actual type parameter belongs in a data object not the DDS for some nuanced reasons around enabling flexibility for handling compatibility errors. Also such typing really should be parried with persisting type information in the document to validate you don't get compatibility errors on accident: SharedTree also does this.

So in short, yes we should aim for that, but its persisted data so its complicated and needs a good compatibility and schema migration story. I'll make sure tree delivers that in a way that works cleanly with container schema, then maybe we can cover the other DDSes.
